x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java

```diff
@@ -0,0 +1,148 @@
+        super(KibanaEnrollmentAction.NAME, transportService, actionFilters, KibanaEnrollmentRequest::new);
+        this.environment = environment;
+        // Should we use a specific origin for this ? Are we satisfied with the auditability of the change password request as-is ?
+        this.client = new OriginSettingClient(client, SECURITY_ORIGIN);
+        this.settings = settings;
+        this.sslService = sslService;
+    }
+
+    @Override protected void doExecute(
+        Task task, KibanaEnrollmentRequest request, ActionListener<KibanaEnrollmentResponse> listener) {
+
+        final KeyConfig keyConfig = sslService.getHttpTransportSSLConfiguration().keyConfig();
+        if (keyConfig instanceof StoreKeyConfig == false) {
+            listener.onFailure(new ElasticsearchException(
+                "Unable to enroll client. Elasticsearch node HTTP layer SSL configuration is not configured with a keystore"));
+            return;
+        }
+        List<X509Certificate> caCertificates;
+        try {
+            caCertificates = ((StoreKeyConfig) keyConfig).x509Certificates(environment)
+                .stream()
+                .filter(x509Certificate -> x509Certificate.getBasicConstraints() != -1)
+                .collect(Collectors.toList());
+        } catch (Exception e) {
+            listener.onFailure(new ElasticsearchException("Unable to enroll client. Cannot retrieve CA certificate for the HTTP layer" +
+                "of the Elasticsearch node.", e));
+            return;
+        }
+        if (caCertificates.size() != 1) {
+            listener.onFailure(new ElasticsearchException(
+                "Unable to enroll client. Elasticsearch node HTTP layer SSL configuration Keystore [xpack.security.http.ssl.keystore] " +
+                    "doesn't contain a single PrivateKey entry where the associated certificate is a CA certificate"));
+
+        } else {
+            String httpCa;
+            try {
+                httpCa = Base64.getUrlEncoder().encodeToString(caCertificates.get(0).getEncoded());
+            } catch (CertificateEncodingException cee) {
+                listener.onFailure(new ElasticsearchException(
+                    "Unable to enroll client. Elasticsearch node HTTP layer SSL configuration uses a malformed CA certificate", cee));
+                return;
+            }
+            final NodesInfoRequest nodesInfoRequest = new NodesInfoRequest().addMetric(NodesInfoRequest.Metric.HTTP.metricName());
+            client.execute(NodesInfoAction.INSTANCE, nodesInfoRequest, ActionListener.wrap(nodesInfoResponse -> {
+                final List<String> nodeList = new ArrayList<>();
+                for (NodeInfo nodeInfo : nodesInfoResponse.getNodes()) {
+                    nodeList.add(nodeInfo.getInfo(HttpInfo.class).getAddress().publishAddress().toString());
+                }
+                // Should we guard against setting the password of the kibana_system user multiple times?
```

Comment:
Until we get an update that service accounts can be used in Kibana, I'm working with the assumption that we'll use a builtin user